src/commands/upgrade-project.ts

```diff
@@ -0,0 +1,137 @@
+      if (jdata.type === "finalize") {
+        const target = jdata.data.project.target;
+        const curKernel = jdata.data.project.templates.find(
+          (t: any) => t.name === "kernel"
+        ).version;
+        const curOkapi = jdata.data.project.templates.find(
+          (t: any) => t.name === "okapilib"
+        ).version;
+        return { target, curKernel, curOkapi };
+      }
+    }
+  }
+  return { target: "", curKernel: "", curOkapi: "" };
+};
+
+/**
+ * Queries the server for the latest available library versions.
+ *
+ * @param target The target device for this project
+ * @returns The kernel and okapi (if applicable) library versions
+ */
+const fetchServerVersions = async (
+  target: string
+): Promise<{ newKernel: string; newOkapi: string | undefined }> => {
+  const { stdout, stderr } = await promisify(child_process.exec)(
+    `pros c q --target ${target} --machine-output`
+  );
+
+  let newKernel = "0.0.0";
+  let newOkapi = "0.0.0";
+
+  for (let e of stdout.split(/\r?\n/)) {
+    if (e.startsWith(PREFIX)) {
+      let jdata = JSON.parse(e.substr(PREFIX.length));
+      if (jdata.type === "finalize") {
+        for (let ver of jdata.data) {
+          if (ver.name === "kernel" && gt(ver.version, newKernel)) {
+            newKernel = ver.version;
+          } else if (ver.name === "okapilib" && gt(ver.version, newOkapi)) {
+            newOkapi = ver.version;
+          }
+        }
+      }
+    }
+  }
+
+  return { newKernel, newOkapi };
+};
+
+/**
+ * Actually performs the upgrade to the latest library versions for the project.
+ */
+const runUpgrade = async () => {
+  const { stdout, stderr } = await promisify(child_process.exec)(
+    `pros c u --project ${vscode.workspace.workspaceFolders?.[0].uri.path} --machine-output`
+  );
+
+  const errorMessage = parseErrorMessage(stdout);
+  if (errorMessage) {
+    throw new Error(errorMessage);
+  }
+};
+
+/**
+ * Confirms with the user that the project should be updated to the specified library versions.
+ *
+ * @param kernel The new kernel version
+ * @param okapi The new Okapilib version
+ */
+const userApproval = async (
+  kernel: string | undefined,
+  okapi: string | undefined
+) => {
+  let title;
+  if (kernel && okapi) {
+    title = `Upgrade to kernel ${kernel} and Okapilib ${okapi}?`;
+  } else if (kernel) {
+    title = `Upgrade to Okapilib ${okapi}?`;
+  } else {
+    title = `Upgrade to kernel ${kernel}?`;
+  }
+  await vscode.window.showQuickPick(
+    [{ label: "yes", description: "recommended" }, { label: "no" }],
+    {
+      placeHolder: "yes",
+      canPickMany: false,
+      title: title,
+    }
+  );
+};
+
+export const upgradeProject = async () => {
+  try {
+    const { target, curKernel, curOkapi } = await fetchTarget();
+    const { newKernel, newOkapi } = await fetchServerVersions(target);
+    if (curKernel === newKernel && curOkapi === newOkapi) {
+      await vscode.window.showInformationMessage("Project is up to date!");
```

Comment:
On a new project that is already up to date I am still being shown the upgrade prompt:

https://user-images.githubusercontent.com/22580992/119344969-4bd70180-bc66-11eb-8ca7-4e48fc2804cf.mov

The prompt is recommending a older version of the kernel and the error is showing a different version as well.
When clicking out of the upgrade prompt is still displays the error message:

https://user-images.githubusercontent.com/22580992/119345252-a2dcd680-bc66-11eb-8689-89a4b3f00e01.mov

Notice the prompt is now recommending a different version of the kernel. Selecting no on the upgrade prompt also still displays the error.

Reply:
This is going to be the case until I fix the CLI bug that reports the wrong available libraries, which I need to put in an issue for. The logic on the VSCode is good afaik but the CLI needs a fix for this bug